src/mlpack/core/optimizers/ada_grad/ada_grad.hpp

```diff
@@ -0,0 +1,172 @@
+   *        function is visited in linear order.
+   */
+  AdaGrad(DecomposableFunctionType& function,
+      const double stepSize = 0.01,
+      const double epsilon = 1e-8,
+      const size_t maxIterations = 100000,
+      const double tolerance = 1e-5,
+      const bool shuffle = true)
+  {
+    AdaGradUpdate adagradUpdate(epsilon);
+    optimizer = SGD<DecomposableFunctionType, AdaGradUpdate>(function, stepSize, maxIterations,
+        tolerance, shuffle, adagradUpdate);  
+  }
+
+  /**
+   * Optimize the given function using AdaGrad. The given starting point will
+   * be modified to store the finishing point of the algorithm, and the final
+   * objective value is returned.
+   *
+   * @param iterate Starting point (will be modified).
+   * @return Objective value of the final point.
+   */
+  double Optimize(arma::mat& iterate)
+  {
+    return optimizer.Optimize(iterate);
+  }
+
+  //! Get the instantiated function to be optimized.
+  const DecomposableFunctionType& Function() const { return function; }
+  //! Modify the instantiated function.
+  DecomposableFunctionType& Function() { return function; }
+
+  //! Get the step size.
+  double StepSize() const { return optimizer.StepSize(); }
+  //! Modify the step size.
+  double& StepSize() { return optimizer.StepSize(); }
+
+  //! Get the value used to initialise the squared gradient parameter.
+  double Epsilon() const { return optimizer.AdaGradUpdate.Epsilon(); }
+  //! Modify the value used to initialise the squared gradient parameter.
+  double& Epsilon() { return optimizer.AdaGradUpdate.Epsilon(); }
+
+  //! Get the maximum number of iterations (0 indicates no limit).
+  size_t MaxIterations() const { return optimizer.MaxIterations(); }
+  //! Modify the maximum number of iterations (0 indicates no limit).
+  size_t& MaxIterations() { return optimizer.MaxIterations(); }
+
+  //! Get the tolerance for termination.
+  double Tolerance() const { return optimizer.Tolerance(); }
+  //! Modify the tolerance for termination.
+  double& Tolerance() { return optimizer.Tolerance(); }
+
+  //! Get whether or not the individual functions are shuffled.
+  bool Shuffle() const { return optimizer.Shuffle(); }
+  //! Modify whether or not the individual functions are shuffled.
+  bool& Shuffle() { return optimizer.Shuffle(); }
+
+ private:
+  //! The instantiated function.
+  DecomposableFunctionType& function;
+
+  //! The step size for each example.
+  double stepSize;
+
+  //! The value used to initialise the squared gradient parameter.
+  double epsilon;
+
+  //! The maximum number of allowed iterations.
+  size_t maxIterations;
+
+  //! The tolerance for termination.
+  double tolerance;
+
+  //! Controls whether or not the individual functions are shuffled when
+  //! iterating.
+  bool shuffle;
+
+  //! Stochastic Gradient Descent object with AdaGrad policy.
+  SGD<DecomposableFunctionType, AdaGradUpdate>
+  optimizer(DecomposableFunctionType&,
```

Comment:
Yes, I have made the change. I was just testing a few things to resolve build errors. 

Actually, I am not able to build locally now as CMake gets stuck after executing few steps when I run command `cmake ..` (details [here](https://paste.debian.net/hidden/e457f447/)). Earlier when I opened up AdaGrad PR, I was able to build and test. Could you give an idea why it might be happening?